tools/src/tensil/tools/compiler/ArenaMemorySpace.scala

```diff
@@ -0,0 +1,52 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright © 2019-2022 Tensil AI Company */
+
+package tensil.tools.compiler
+
+import tensil.tools.CompilerException
+
+object ArenaMemorySpace {
+  def apply(
+      name: String,
+      tag: MemoryTag,
+      depth: MemoryAddressRaw
+  ) =
+    new ArenaMemorySpace(
+      name,
+      tag,
+      depth,
+      MemoryAddressRaw.Zero
+    )
+}
+
+class ArenaMemorySpace private (
+    val name: String,
+    val tag: MemoryTag,
+    val depth: MemoryAddressRaw,
+    private var next: MemoryAddressRaw
+) extends MemorySpace {
+  def allocate(
+      ref: MemoryRef,
+      size: MemoryAddressRaw
+  ): Option[MemorySpan] =
+    if (next + size <= depth) {
+      val span = MemoryAddress.mkSpan(
+        tag,
+        ref,
+        next,
+        size
+      )
+
+      next += size
+
+      Some(span)
+    } else None
+
+  def free(span: MemorySpan): Unit =
+    throw new CompilerException(
+      "Freeing from arena memory space is not supported"
```

Comment:
Just curious, why can't we free from arena?

Reply:
`ArenaMemorySpace` is used in cases where the entire memory space is being freed at once, such as `IsolatedLocalScheduler`. Since freeing individual spans is not necessary the `ArenaMemorySpace` implementation can be very simple.